src/singledocparser.cpp

```diff
@@ -46,6 +46,9 @@ void SingleDocParser::HandleDocument(EventHandler& eventHandler) {
 }
 
 void SingleDocParser::HandleNode(EventHandler& eventHandler) {
+  if (depth > depth_limit) {
+    throw ParserException(m_scanner.mark(), ErrorMsg::BAD_FILE);
```

Comment:
> Honestly, I find the types you suggested to be a little on the heavy side. Isn't there something in the standard library or boost that could cover for this without having to reinvent the wheel?

That's a great question. I'm not familiar with a type _with this exact purpose_ from either the standard library or the boost libraries. Is it really reinventing the wheel then? If you feel strongly about it, the boost mailing list should be able to point you in the direction of people with more experience :) Having been to cppcon, I can tell you that a lot of different people are _really_ responsive to useful stuff.

Assuming it's not already a thing, I would guess that it's for the same reason that there's no standard process library (actually... a week ago, `boost::process` was released): the standard focuses on portability and stacks aren't actually a fully portable concept.

> I like the original implementation I provided because it's just a single integer to keep track of, without that data structure around..

The data structure has no virtual table and stores a simple reference to plain old data. In debug builds, the internal reference _may_ end up being a pointer allocation (on the stack). In any sort of reasonable optimization, its code should get inlined. The optimizer should also identify that the reference is redundant to `this->depth` and _may or may not_ elect to remove the pointer (since `depth` is not marked volatile it becomes whether or not the optimizer chooses to use a single indirection (at the cost of a pointer) or double indirection (and save the bytes); on x86 it'll typically choose the former because the second indirection is constant (offset of `depth`) and can be directly instrumented in assembly). and the class basically gets completely removed from the stack. It then becomes just a source code concept to tell the compiler how to do things with zero cost to runtime. This is why I don't think it's heavy.

I think you're confusing `verbose` with `heavy`, which is arguable: when you consider the limited scope of this one specific function, sure, it's rather verbose.

Wow that's a lot of words. Let me just tl;dr that for you:
[Optimization level 3 assembly with `depth_guard` class](https://godbolt.org/g/0NVBQT)
[Optimization level 3 assembly with equivalent manual control](https://godbolt.org/g/FVVCmE)

IN particular, the `depth_guard` assembly is both shorter and generally simpler. Importantly, the function prolog (setting up the stack frame, et al) is identical which means that `stack_guard` definition gets elided and optimized away.

Although I'm not quite sure what's up with the repeated `sub eax, 1` in .L8.